src/lib/modules/specialconfigs/index.js

```diff
@@ -162,25 +185,101 @@ class SpecialConfigs {
       let cmd = params.contract.deployIf;
+      const contract = params.contract;
       if (!cmd) {
         return cb(params);
       }
 
-      self.events.request('runcode:eval', cmd, (err, result) => {
+      if (typeof cmd === 'function') {
+        this.getOnDeployLifecycleHookDependencies(contract).then(dependencies => {
+          const callbackFn = (value) => {
+          };
+          (async () => {
+            params.shouldDeploy = await contract.deployIf(dependencies);
+            cb(params);
+          })();
+        }, err => {
+          return cb(new Error(`Error when registering deployIf hook for ${contract.name}: ${err.message}`));
+        });
+      } else {
+
+        self.events.request('runcode:eval', cmd, (err, result) => {
+          if (err) {
+            self.logger.error(params.contract.className + ' deployIf directive has an error; contract will not deploy');
+            self.logger.error(err);
+            params.shouldDeploy = false;
+          } else if (!result) {
+            self.logger.info(params.contract.className + ' deployIf directive returned false; contract will not deploy');
+            params.shouldDeploy = false;
+          }
+
+          cb(params);
+        });
+      }
+    });
+  }
+
+  getOnDeployLifecycleHookDependencies(contractConfig) {
+    let dependencyNames = contractConfig.deps || [];
+    dependencyNames.push(contractConfig.className);
+
+    return new Promise((resolve, reject) => {
+      async.map(dependencyNames, (contractName, next) => {
+        this.events.request('contracts:contract', contractName, (contractRecipe) => {
+          if (!contractRecipe) {
+            next(new Error('ReferredContractDoesNotExist'));
+          }
+          this.events.request('blockchain:contract:create', {
+            abi: contractRecipe.abiDefinition,
+            address: contractRecipe.deployedAddress
+          }, contractInstance => {
+            next(null, { className: contractRecipe.className, instance: contractInstance });
+          });
+        });
+      }, (err, contractInstances) => {
         if (err) {
-          self.logger.error(params.contract.className + ' deployIf directive has an error; contract will not deploy');
-          self.logger.error(err);
-          params.shouldDeploy = false;
-        } else if (!result) {
-          self.logger.info(params.contract.className + ' deployIf directive returned false; contract will not deploy');
-          params.shouldDeploy = false;
+          reject(err);
         }
+        this.events.request('blockchain:get', web3 => {
+
+          let dependencies = { contracts: {}, web3 };
 
-        cb(params);
+          contractInstances.forEach(contractInstance => {
+            dependencies.contracts[contractInstance.className] = contractInstance.instance;
+          });
+          resolve(dependencies);
+        });
       });
     });
   }
 
+  getAfterDeployLifecycleHookDependencies() {
```

Comment:
`getOnDeployLifecycleHookDependencies` and `getAfterDeployLifecycleHookDependencies` seems to share some logic, do you think it could be possible to extract the sharing logic into function?

Reply:
Yes I was considering that too. Especially the part that puts together the dependencies ( think the rest is already pretty broken down into pieces and now really reusable).

But then I thought, it might be a little overkill to introduce a new function for:

```
          this.events.request('blockchain:get', web3 => {

            let dependencies = { contracts: {}, web3 };

            contractInstances.forEach(contractInstance => {
              dependencies.contracts[contractInstance.className] = contractInstance.instance;
            });
            resolve(dependencies);
          });
```

Anyways, will update it :)

Reply:
To reduce the shared code and also not having to re-create the contract object every time there is a new onDeploy, I think we should have one of the contract modules have a list of the web3 contracts an retrieve them here with an event.

Every time a contract is deployed, it would create that contract object. That would be quite useful for a couple of our modules. I'm thinking about this one, the ENS one, maybe even tests.

Maybe one of our modules already does that, but I don't recall. Also, maybe there is a good reason we didn't do that before (maybe a memory issue)? What do you think?

Reply:
This is a very good point you're raising here @jrainville 

From the top of my head I'd say `ContractsManager` should be responsible for that. re: Memory issues: AFAIK node has a max memory limit of ~1.5 GB by default but can be easily extended if needed.

I guess that's something we'll have to experiment with.

I believe this would be a good candidate for a separate PR/effort though.